### PR TITLE
Improve decoration system in ir nodes

### PR DESCRIPTION
**Pull request by jdconrad**
This change creates consistency in the decorations added for the ir nodes by adding addtional methods getValue and toString to the base IRDecoration class. This also add getDecorationValue and getDecorationString as convenience methods to IRNode where getDecoration(IRDExpressionType.class).getType() becomes getDecorationValue(IRDExpressionType.class). The BinaryMathNode is used an example of conversion to the new methods. The rest of the nodes will change in a follow up.

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/bwc

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java [hunk 1]
```diff
@@ -31,40 +31,56 @@
 
     /* ---- begin decorations ---- */
 
-    public interface IRDecoration {
+    public abstract static class IRDecoration<V> {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java:34]: This is a substantional enough class that it should be pulled out into it's own file.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java [hunk 2]
```diff
@@ -31,40 +31,56 @@
...
 
-        if (decoration != null) {
-            attachDecoration(decoration);
+    public <T extends IRDecoration<V>, V> V getDecorationValue(Class<T> type) {
+        return getDecorationValueOrDefaultValue(type, null);
+    }
 
-            return true;
-        }
+    public <T extends IRDecoration<V>, V> V getDecorationValueOrDefaultValue(Class<T> type, V defaultValue) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java:76]: `getDecorationValueOrDefault`

**Review by stu-elastic — APPROVED** [review]
lgtm

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java:34]: I agree, but I'm going to hold off on this for now as I think we should reorg everything into user and ir packages with subpackages for node and phase. I will do that in a follow up PR.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/IRNode.java]: Done.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review! Will commit as soon as CI passes.

**Pull request merged**